API Design on unstable tcp server: purpose of NewService trait

Hello,

I'm running into some API ergonomic issues when trying to use the TCP server module part of this crate. I have attempted to extend the example `tcp-server.rs` with a new method: 

```rust
impl MbServer {
    async fn run(self, server: server::tcp::Server) -> std::io::Result<()> {
        server.serve(self).await
    }
}
```

The first issue I run into is this compiler error:
```
error[E0277]: expected a `Fn<()>` closure, found `MbServer`
  --> examples/tcp-server.rs:31:22
   |
31 |         server.serve(self).await
   |                ----- ^^^^ expected an `Fn<()>` closure, found `MbServer`
   |                |
   |                required by a bound introduced by this call
   |
   = help: the trait `Fn<()>` is not implemented for `MbServer`
   = note: wrap the `MbServer` in a closure with no arguments: `|| { /* code */ }`
   = note: required because of the requirements on the impl of `NewService` for `MbServer`
```
Fair enough, the architecture is built around the idea of a "callable" service. So I refactored to:
```rust
server.serve(|| Ok(self)).await
```
But then it gets very much in the way:
```
error[E0525]: expected a closure that implements the `Fn` trait, but this closure only implements `FnOnce`
  --> examples/tcp-server.rs:31:22
   |
31 |         server.serve(|| Ok(self)).await
   |                ----- ^^^^^^----^
   |                |     |     |
   |                |     |     closure is `FnOnce` because it moves the variable `self` out of its environment
   |                |     this closure implements `FnOnce`, not `Fn`
   |                the requirement to implement `Fn` derives from here
```
As I'm moving the struct implementing Service here, this restriction on the API seems unnecessary.

While digging to figure out why the API requires this, I stumbled on the `NewService` trait and it's not clear to me what it's role was. I adjusted the functionality of `serve` et al to bound `S` as `Service` instead of `NewService` and my original implementation compiled fine and functioned as I expected.

It seems a lot of complexity and extra restrictions are added as a result of this `NewService` trait with no obvious benefit. Perhaps I'm misunderstanding the use case, but as the API is unstable, would it perhaps make sense to remove it or add more functions for addressing the extra complexity of the other use case only when necessary?

--- a/.github/workflows/continuous-integration.yaml
+++ b/.github/workflows/continuous-integration.yaml
@@ -39,7 +39,7 @@ jobs:
 
     steps:
       - name: Install Rust toolchain
-        uses: dtolnay/rust-toolchain@master
+        uses: dtolnay/rust-toolchain@v1
         with:
           toolchain: stable
           targets: ${{ matrix.target }}

--- a/.github/workflows/pre-commit.yaml
+++ b/.github/workflows/pre-commit.yaml
@@ -29,7 +29,7 @@ jobs:
           python-version: "3.x"
 
       - name: Install Rust toolchain
-        uses: dtolnay/rust-toolchain@master
+        uses: dtolnay/rust-toolchain@v1
         with:
           toolchain: stable
           components: clippy, rustfmt

--- a/.github/workflows/security-audit.yaml
+++ b/.github/workflows/security-audit.yaml
@@ -20,7 +20,7 @@ jobs:
   security_audit:
     runs-on: ubuntu-latest
     steps:
-      - uses: dtolnay/rust-toolchain@master
+      - uses: dtolnay/rust-toolchain@v1
         with:
           toolchain: stable
 

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -49,7 +49,7 @@ repos:
       - id: check-github-actions
       - id: check-github-workflows
   - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: v3.0.0-alpha.5
+    rev: v3.0.0-alpha.6
     hooks:
       - id: prettier
         types_or:

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -49,7 +49,7 @@ repos:
       - id: check-github-actions
       - id: check-github-workflows
   - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: v3.0.0-alpha.4
+    rev: v3.0.0-alpha.5
     hooks:
       - id: prettier
         types_or:

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,10 +3,16 @@
 
 # Changelog
 
+## v0.8.0 (Unreleased)
+
+- TCP Server: Replace `NewService` trait with closures and never panic
+- TCP Server: Stabilize "tcp-server" feature
+- RTU Server: Remove `NewService` trait and pass instance directly
+
 ## v0.7.1 (2023-02-28)
 
 - Fix (sync): Panic when using client-side timeouts [#155](https://github.com/slowtec/tokio-modbus/issues/155)
-- tcp-server-unstable: Upgrade socket2 dependency
+- tcp-server: Upgrade socket2 dependency
 
 ## v0.7.0 (2023-02-14)
 
@@ -47,7 +53,7 @@
 ## v0.5.1 (2021-11-21)
 
 - Fix: require tokio/rt for sync feature
-- Changed: Update methods on TCP server to be async (only concerns `tcp-server-unstable` feature)
+- Changed: Update methods on TCP server to be async (only concerns `tcp-server` feature)
 
 ## v0.5.0 (2021-08-20)
 
@@ -73,7 +79,7 @@
 - Make `Exception` and `ExceptionResponse` public
 - Fixed `WriteSingleCoil` response to include data
 - Hide server traits `Service`/`NewService` traits behind `server` feature
-- Hide TCP server implementation behind `tcp-server-unstable` feature
+- Hide TCP server implementation behind `tcp-server` feature
 - Improved documentation
 
 ### Breaking Changes

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -23,7 +23,7 @@ all-features = true
 # <https://users.rust-lang.org/t/psa-please-specify-precise-dependency-versions-in-cargo-toml/71277>
 
 [dependencies]
-async-trait = "0.1.64"
+async-trait = "0.1.65"
 byteorder = "1.4.3"
 bytes = "1.4.0"
 futures = { version = "0.3.26", optional = true }
@@ -31,7 +31,7 @@ futures-util = { version = "0.3.26", optional = true, default-features = false }
 log = "0.4.17"
 smallvec = { version = "1.10.0", default-features = false }
 socket2 = { version = "0.5.1", optional = true, default-features = false }
-tokio = { version = "1.25.0", default-features = false }
+tokio = { version = "1.26.0", default-features = false }
 # Disable default-features to exclude unused dependency on libudev
 tokio-serial = { version = "5.4.4", optional = true, default-features = false }
 tokio-util = { version = "0.7.7", default-features = false, features = [
@@ -39,9 +39,10 @@ tokio-util = { version = "0.7.7", default-features = false, features = [
 ] }
 
 [dev-dependencies]
+anyhow = "1.0.69"
 env_logger = "0.10.0"
 futures = "0.3.26"
-tokio = { version = "1.25.0", default-features = false, features = [
+tokio = { version = "1.26.0", default-features = false, features = [
     "macros",
     "rt-multi-thread",
     "time",

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -4,7 +4,7 @@
 [package]
 name = "tokio-modbus"
 description = "Tokio-based Modbus library"
-version = "0.7.1"
+version = "0.8.0"
 authors = [
     "slowtec GmbH <post@slowtec.de>",
     "Markus Kohlhase <markus.kohlhase@slowtec.de>",
@@ -50,13 +50,13 @@ tokio = { version = "1.26.0", default-features = false, features = [
 tokio-serial = { version = "5.4.4", default-features = false }
 
 [features]
-default = ["rtu", "tcp"]
+default = ["rtu", "tcp", "rtu-server", "tcp-server"]
 rtu = ["futures-util/sink"]
 tcp = ["tokio/net", "futures-util/sink"]
 rtu-sync = ["rtu", "sync", "dep:tokio-serial"]
 tcp-sync = ["tcp", "sync"]
 rtu-server = ["rtu", "server", "tokio/macros", "dep:tokio-serial"]
-tcp-server-unstable = [
+tcp-server = [
     "tcp",
     "server",
     "socket2/all",
@@ -108,4 +108,4 @@ required-features = ["tcp"]
 [[example]]
 name = "tcp-server"
 path = "examples/tcp-server.rs"
-required-features = ["tcp-server-unstable"]
+required-features = ["tcp-server"]

--- a/README.md
+++ b/README.md
@@ -97,9 +97,9 @@ cargo test --workspace --all-features
 
 ## Protocol-Specification
 
-- [MODBUS Application Protocol Specification v1.1b3 (PDF)](http://modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf)
-- [MODBUS over serial line specification and implementation guide v1.02 (PDF)](http://modbus.org/docs/Modbus_over_serial_line_V1_02.pdf)
-- [MODBUS Messaging on TCP/IP Implementation Guide v1.0b (PDF)](http://modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf)
+- [Modbus Application Protocol Specification v1.1b3 (PDF)](http://modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf)
+- [Modbus over serial line specification and implementation guide v1.02 (PDF)](http://modbus.org/docs/Modbus_over_serial_line_V1_02.pdf)
+- [Modbus Messaging on TCP/IP Implementation Guide v1.0b (PDF)](http://modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf)
 
 ## License
 

--- a/README.md
+++ b/README.md
@@ -46,7 +46,7 @@ tokio-modbus = "*"
 - `"rtu-sync`: Synchronous RTU client
 - `"tcp-sync"`: Synchronous TCP client
 - `"rtu-server"`: (Asynchronous) RTU server
-- `"tcp-server-unstable"`: (Asynchronous) TCP server (experimental)
+- `"tcp-server"`: (Asynchronous) TCP server
 
 #### Examples
 
@@ -75,7 +75,7 @@ For a TCP server:
 
 ```toml
 [dependencies]
-tokio-modbus = { version = "*", default-features = false, features = ["tcp-server-unstable"] }
+tokio-modbus = { version = "*", default-features = false, features = ["tcp-server"] }
 ```
 
 ## Examples

--- a/examples/rtu-server-address.rs
+++ b/examples/rtu-server-address.rs
@@ -7,14 +7,13 @@ use std::{thread, time::Duration};
 
 use futures::future;
 
-use tokio_modbus::prelude::*;
-use tokio_modbus::server::{self, Service};
+use tokio_modbus::{prelude::*, server::rtu::Server};
 
-struct Server {
+struct Service {
     slave: Slave,
 }
 
-impl Service for Server {
+impl tokio_modbus::server::Service for Service {
     type Request = SlaveRequest;
     type Response = Option<Response>;
     type Error = std::io::Error;
@@ -44,9 +43,12 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     println!("Starting up server...");
     let _server = thread::spawn(move || {
         let rt = tokio::runtime::Runtime::new().unwrap();
-        let server = server::rtu::Server::new(server_serial);
+        let server = Server::new(server_serial);
+        let service = Service { slave };
         rt.block_on(async {
-            server.serve_forever(move || Ok(Server { slave })).await;
+            if let Err(err) = server.serve_forever(service).await {
+                eprintln!("{err}");
+            }
         });
     });
 

--- a/examples/rtu-server.rs
+++ b/examples/rtu-server.rs
@@ -7,19 +7,18 @@ use std::{thread, time::Duration};
 
 use futures::future;
 
-use tokio_modbus::prelude::*;
-use tokio_modbus::server::{self, Service};
+use tokio_modbus::{prelude::*, server::rtu::Server};
 
-struct Server;
+struct Service;
 
-impl Service for Server {
-    type Request = Request;
+impl tokio_modbus::server::Service for Service {
+    type Request = SlaveRequest;
     type Response = Response;
     type Error = std::io::Error;
     type Future = future::Ready<Result<Self::Response, Self::Error>>;
 
     fn call(&self, req: Self::Request) -> Self::Future {
-        match req {
+        match req.request {
             Request::ReadInputRegisters(_addr, cnt) => {
                 let mut registers = vec![0; cnt.into()];
                 registers[2] = 0x77;
@@ -38,9 +37,12 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     println!("Starting up server...");
     let _server = thread::spawn(move || {
         let rt = tokio::runtime::Runtime::new().unwrap();
-        let server = server::rtu::Server::new(server_serial);
+        let server = Server::new(server_serial);
+        let service = Service;
         rt.block_on(async {
-            server.serve_forever(|| Ok(Server)).await;
+            if let Err(err) = server.serve_forever(service).await {
+                eprintln!("{err}");
+            }
         });
     });
 

--- a/examples/tcp-server.rs
+++ b/examples/tcp-server.rs
@@ -3,15 +3,16 @@
 
 //! TCP server example
 
-use futures::future;
 use std::{net::SocketAddr, time::Duration};
 
-use tokio_modbus::prelude::*;
-use tokio_modbus::server::{self, Service};
+use futures::future;
+use tokio::net::TcpListener;
 
-struct MbServer;
+use tokio_modbus::{prelude::*, server::tcp::Server};
 
-impl Service for MbServer {
+struct Service;
+
+impl tokio_modbus::server::Service for Service {
     type Request = Request;
     type Response = Response;
     type Error = std::io::Error;
@@ -41,10 +42,16 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     Ok(())
 }
 
-async fn server_context(socket_addr: SocketAddr) {
-    println!("Starting up server...");
-    let server = server::tcp::Server::new(socket_addr);
-    server.serve(|| Ok(MbServer)).await.unwrap();
+async fn server_context(socket_addr: SocketAddr) -> anyhow::Result<()> {
+    println!("Starting up server on {socket_addr}");
+    let listener = TcpListener::bind(socket_addr).await?;
+    let server = Server::new(listener);
+    let new_service = |_socket_addr| Some(Service);
+    let on_process_error = |err| {
+        eprintln!("{err}");
+    };
+    server.serve(&new_service, on_process_error).await?;
+    Ok(())
 }
 
 async fn client_context(socket_addr: SocketAddr) {

--- a/src/codec/rtu.rs
+++ b/src/codec/rtu.rs
@@ -11,8 +11,8 @@ use smallvec::SmallVec;
 use std::io::{Cursor, Error, ErrorKind, Result};
 use tokio_util::codec::{Decoder, Encoder};
 
-// [MODBUS over Serial Line Specification and Implementation Guide V1.02](http://modbus.org/docs/Modbus_over_serial_line_V1_02.pdf), page 13
-// "The maximum size of a MODBUS RTU frame is 256 bytes."
+// [Modbus over Serial Line Specification and Implementation Guide V1.02](http://modbus.org/docs/Modbus_over_serial_line_V1_02.pdf), page 13
+// "The maximum size of a Modbus RTU frame is 256 bytes."
 const MAX_FRAME_LEN: usize = 256;
 
 type DroppedBytes = SmallVec<[u8; MAX_FRAME_LEN]>;

--- a/src/server/mod.rs
+++ b/src/server/mod.rs
@@ -9,9 +9,18 @@
 #[cfg(feature = "rtu-server")]
 pub mod rtu;
 
-#[cfg(feature = "tcp-server-unstable")]
+#[cfg(feature = "tcp-server")]
 pub mod tcp;
 
 mod service;
+pub use self::service::Service;
 
-pub use service::{NewService, Service};
+/// Cause for termination
+#[derive(Debug, Clone)]
+pub enum Terminated {
+    /// The server has finished processing.
+    Finished,
+
+    /// Processing has been aborted.
+    Aborted,
+}

--- a/src/server/rtu.rs
+++ b/src/server/rtu.rs
@@ -3,16 +3,23 @@
 
 //! Modbus RTU server skeleton
 
-use crate::{
-    codec,
-    frame::*,
-    server::service::{NewService, Service},
-};
-use futures::{select, Future, FutureExt as _};
+use std::{io, path::Path};
+
+use futures::{Future, FutureExt as _};
 use futures_util::{SinkExt as _, StreamExt as _};
-use std::{io::Error, path::Path};
 use tokio_serial::SerialStream;
 use tokio_util::codec::Framed;
+
+use crate::{
+    codec::rtu::ServerCodec,
+    frame::{
+        rtu::{RequestAdu, ResponseAdu},
+        OptionalResponsePdu,
+    },
+    server::service::Service,
+};
+
+use super::Terminated;
 
 #[derive(Debug)]
 pub struct Server {
@@ -21,7 +28,7 @@ pub struct Server {
 
 impl Server {
     /// set up a new Server instance from an interface path and baud rate
-    pub fn new_from_path<P: AsRef<Path>>(p: P, baud_rate: u32) -> Result<Self, Error> {
+    pub fn new_from_path<P: AsRef<Path>>(p: P, baud_rate: u32) -> io::Result<Self> {
         let serial =
             SerialStream::open(&tokio_serial::new(p.as_ref().to_string_lossy(), baud_rate))?;
         Ok(Server { serial })
@@ -33,81 +40,76 @@ impl Server {
         Server { serial }
     }
 
-    /// serve Modbus RTU requests based on the provided service until it finishes
-    pub async fn serve_forever<S, Req, Res>(self, new_service: S)
+    /// Process Modbus RTU requests.
+    pub async fn serve_forever<S>(self, service: S) -> io::Result<()>
     where
-        S: NewService<Request = Req, Response = Res> + Send + Sync + 'static,
-        Req: From<rtu::RequestAdu> + Send,
-        Res: Into<OptionalResponsePdu> + Send,
-        S::Instance: Send + Sync + 'static,
-        S::Error: Into<Error>,
+        S: Service + Send + Sync + 'static,
+        S::Request: From<RequestAdu> + Send,
+        S::Response: Into<OptionalResponsePdu> + Send,
+        S::Error: Into<io::Error>,
     {
-        self.serve_until(new_service, futures::future::pending())
-            .await;
+        let framed = Framed::new(self.serial, ServerCodec::default());
+        process(framed, service).await
     }
 
-    /// serve Modbus RTU requests based on the provided service until it finishes or a shutdown signal is received
-    pub async fn serve_until<S, Req, Res, Sd>(self, new_service: S, shutdown_signal: Sd)
+    /// Process Modbus RTU requests until finished or aborted.
+    ///
+    /// Warning: Request processing is not scoped and could be aborted at any internal await point!
+    /// See also: <https://rust-lang.github.io/wg-async/vision/roadmap/scopes.html#cancellation>
+    pub async fn serve_until<S, X>(self, service: S, abort_signal: X) -> io::Result<Terminated>
     where
-        S: NewService<Request = Req, Response = Res> + Send + Sync + 'static,
-        Sd: Future<Output = ()> + Sync + Send + Unpin + 'static,
-        Req: From<rtu::RequestAdu> + Send,
-        Res: Into<OptionalResponsePdu> + Send,
-        S::Instance: Send + Sync + 'static,
-        S::Error: Into<Error>,
+        S: Service + Send + Sync + 'static,
+        S::Request: From<RequestAdu> + Send,
+        S::Response: Into<OptionalResponsePdu> + Send,
+        S::Error: Into<io::Error>,
+        X: Future<Output = ()> + Sync + Send + Unpin + 'static,
     {
-        let framed = Framed::new(self.serial, codec::rtu::ServerCodec::default());
-        let service = new_service.new_service().unwrap();
-        let future = process(framed, service);
-
-        let mut server = Box::pin(future).fuse();
-        let mut shutdown = shutdown_signal.fuse();
-
-        async {
-            select! {
-                res = server => if let Err(e) = res {
-                    println!("error: {e}");
-                },
-                _ = shutdown => println!("Shutdown signal received")
+        let framed = Framed::new(self.serial, ServerCodec::default());
+        let abort_signal = abort_signal.fuse();
+        tokio::select! {
+            res = process(framed, service) => {
+                res.map(|()| Terminated::Finished)
+            },
+            () = abort_signal => {
+                Ok(Terminated::Aborted)
             }
         }
-        .await;
     }
 }
 
 /// frame wrapper around the underlying service's responses to forwarded requests
 async fn process<S, Req, Res>(
-    mut framed: Framed<SerialStream, codec::rtu::ServerCodec>,
+    mut framed: Framed<SerialStream, ServerCodec>,
     service: S,
-) -> Result<(), Error>
+) -> io::Result<()>
 where
     S: Service<Request = Req, Response = Res> + Send + Sync + 'static,
-    S::Request: From<rtu::RequestAdu> + Send,
+    S::Request: From<RequestAdu> + Send,
     S::Response: Into<OptionalResponsePdu> + Send,
-    S::Error: Into<Error>,
+    S::Error: Into<io::Error>,
 {
     loop {
-        let request = match framed.next().await {
-            // Stream is exhausted
-            None => break,
-            Some(request) => request,
-        }?;
+        let Some(request) = framed.next().await.transpose()? else {
+            log::debug!("Stream has finished");
+            break;
+        };
 
         let hdr = request.hdr;
-        let response: OptionalResponsePdu = service
+        let OptionalResponsePdu(Some(response_pdu)) = service
             .call(request.into())
             .await
             .map_err(Into::into)?
-            .into();
+            .into() else {
+                log::debug!("Sending no response for request {hdr:?}");
+                continue;
+        };
 
-        match response.0 {
-            Some(pdu) => {
-                framed.send(rtu::ResponseAdu { hdr, pdu }).await?;
-            }
-            None => {
-                log::debug!("No response for request {hdr:?}");
-            }
-        }
+        framed
+            .send(ResponseAdu {
+                hdr,
+                pdu: response_pdu,
+            })
+            .await?;
     }
     Ok(())
 }

--- a/src/server/service.rs
+++ b/src/server/service.rs
@@ -1,7 +1,7 @@
 // SPDX-FileCopyrightText: Copyright (c) 2017-2023 slowtec GmbH <post@slowtec.de>
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
-use std::{future::Future, io, rc::Rc, sync::Arc};
+use std::future::Future;
 
 /// A Modbus server service.
 pub trait Service {
@@ -19,92 +19,4 @@ pub trait Service {
 
     /// Process the request and return the response asynchronously.
     fn call(&self, req: Self::Request) -> Self::Future;
-}
-
-/// Creates new `Service` values.
-pub trait NewService {
-    /// Requests handled by the service
-    type Request;
-
-    /// Responses given by the service
-    type Response;
-
-    /// Errors produced by the service
-    type Error;
-
-    /// The `Service` value created by this factory
-    type Instance: Service<Request = Self::Request, Response = Self::Response, Error = Self::Error>;
-
-    /// Create and return a new service value.
-    fn new_service(&self) -> io::Result<Self::Instance>;
-}
-
-impl<F, R> NewService for F
-where
-    F: Fn() -> io::Result<R>,
-    R: Service,
-{
-    type Request = R::Request;
-    type Response = R::Response;
-    type Error = R::Error;
-    type Instance = R;
-
-    fn new_service(&self) -> io::Result<R> {
-        (*self)()
-    }
-}
-
-impl<S: NewService + ?Sized> NewService for Arc<S> {
-    type Request = S::Request;
-    type Response = S::Response;
-    type Error = S::Error;
-    type Instance = S::Instance;
-
-    fn new_service(&self) -> io::Result<S::Instance> {
-        (**self).new_service()
-    }
-}
-
-impl<S: NewService + ?Sized> NewService for Rc<S> {
-    type Request = S::Request;
-    type Response = S::Response;
-    type Error = S::Error;
-    type Instance = S::Instance;
-
-    fn new_service(&self) -> io::Result<S::Instance> {
-        (**self).new_service()
-    }
-}
-
-impl<S: Service + ?Sized + 'static> Service for Box<S> {
-    type Request = S::Request;
-    type Response = S::Response;
-    type Error = S::Error;
-    type Future = S::Future;
-
-    fn call(&self, request: S::Request) -> Self::Future {
-        (**self).call(request)
-    }
-}
-
-impl<S: Service + ?Sized + 'static> Service for Rc<S> {
-    type Request = S::Request;
-    type Response = S::Response;
-    type Error = S::Error;
-    type Future = S::Future;
-
-    fn call(&self, request: S::Request) -> Self::Future {
-        (**self).call(request)
-    }
-}
-
-impl<S: Service + ?Sized + 'static> Service for Arc<S> {
-    type Request = S::Request;
-    type Response = S::Response;
-    type Error = S::Error;
-    type Future = S::Future;
-
-    fn call(&self, request: S::Request) -> Self::Future {
-        (**self).call(request)
-    }
 }

--- a/src/server/tcp.rs
+++ b/src/server/tcp.rs
@@ -3,135 +3,145 @@
 
 //! Modbus TCP server skeleton
 
-use crate::{
-    codec,
-    frame::*,
-    server::service::{NewService, Service},
-};
+use std::{io, net::SocketAddr};
 
+use async_trait::async_trait;
 use futures::{self, Future};
 use futures_util::{future::FutureExt as _, sink::SinkExt as _, stream::StreamExt as _};
 use socket2::{Domain, Socket, Type};
-use std::{
-    io::{self, Error},
-    net::SocketAddr,
-    sync::Arc,
-};
 use tokio::net::{TcpListener, TcpStream};
 use tokio_util::codec::Framed;
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+use crate::{
+    codec::tcp::ServerCodec,
+    frame::{
+        tcp::{RequestAdu, ResponseAdu},
+        OptionalResponsePdu,
+    },
+    server::service::Service,
+};
+
+use super::Terminated;
+
+#[async_trait]
+pub trait BindSocket {
+    type Error;
+
+    async fn bind_socket(addr: SocketAddr) -> Result<Socket, Self::Error>;
+}
+
+#[derive(Debug)]
 pub struct Server {
-    socket_addr: SocketAddr,
+    listener: TcpListener,
 }
 
 impl Server {
-    /// Set the address for the server (mandatory).
+    /// Attach the Modbus server to a TCP socket server.
     #[must_use]
-    pub fn new(socket_addr: SocketAddr) -> Self {
-        Self { socket_addr }
+    pub fn new(listener: TcpListener) -> Self {
+        Self { listener }
     }
 
-    /// Start an async Modbus TCP server task.
-    pub async fn serve<S, Req, Res>(&self, service: S) -> Result<(), std::io::Error>
+    /// Start a Modbus TCP server task.
+    pub async fn serve<S, OnConnected, OnProcessError>(
+        &self,
+        on_connected: &OnConnected,
+        on_process_error: OnProcessError,
+    ) -> io::Result<()>
     where
-        S: NewService<Request = Req, Response = Res> + Send + Sync + 'static,
-        Req: From<tcp::RequestAdu> + Send,
-        Res: Into<OptionalResponsePdu> + Send,
-        S::Instance: Send + Sync + 'static,
-        S::Error: Into<Error>,
+        S: Service + Send + Sync + 'static,
+        S::Request: From<RequestAdu> + Send,
+        S::Response: Into<OptionalResponsePdu> + Send,
+        S::Error: Into<io::Error>,
+        OnConnected: Fn(SocketAddr) -> Option<S>,
+        OnProcessError: FnOnce(io::Error) + Clone + Send + 'static,
     {
-        let service = Arc::new(service);
-        let listener = TcpListener::bind(self.socket_addr).await?;
-
         loop {
-            let (stream, _) = listener.accept().await?;
-            let framed = Framed::new(stream, codec::tcp::ServerCodec::default());
-            let new_service = service.clone();
+            let (stream, socket_addr) = self.listener.accept().await?;
+            log::debug!("Accepted connection from {socket_addr}");
 
-            tokio::spawn(Box::pin(async move {
-                let service = new_service.new_service().unwrap();
+            let Some(service) = on_connected(socket_addr) else {
+                log::debug!("No service for connection from {socket_addr}");
+                continue;
+            };
+            let on_process_error = on_process_error.clone();
+            let framed = Framed::new(stream, ServerCodec::default());
+
+            tokio::spawn(async move {
+                log::debug!("Processing requests from {socket_addr}");
                 if let Err(err) = process(framed, service).await {
-                    eprintln!("{err:?}");
+                    on_process_error(err);
                 }
-            }));
+            });
         }
     }
 
-    /// Start a Modbus TCP server that blocks the current thread until a shutdown is requested
-    pub fn serve_until<S, Req, Res, Sd>(self, service: S, shutdown_signal: Sd)
+    /// Start an abortable Modbus TCP server task.
+    ///
+    /// Warning: Request processing is not scoped and could be aborted at any internal await point!
+    /// See also: <https://rust-lang.github.io/wg-async/vision/roadmap/scopes.html#cancellation>
+    pub async fn serve_until<S, X, OnConnected, OnProcessError>(
+        self,
+        on_connected: &OnConnected,
+        on_process_error: OnProcessError,
+        abort_signal: X,
+    ) -> io::Result<Terminated>
     where
-        S: NewService<Request = Req, Response = Res> + Send + Sync + 'static,
-        Sd: Future<Output = ()> + Sync + Send + Unpin + 'static,
-        Req: From<tcp::RequestAdu> + Send,
-        Res: Into<OptionalResponsePdu> + Send,
-        S::Instance: Send + Sync + 'static,
-        S::Error: Into<Error>,
+        S: Service + Send + Sync + 'static,
+        S::Request: From<RequestAdu> + Send,
+        S::Response: Into<OptionalResponsePdu> + Send,
+        S::Error: Into<io::Error>,
+        X: Future<Output = ()> + Sync + Send + Unpin + 'static,
+        OnConnected: Fn(SocketAddr) -> Option<S>,
+        OnProcessError: FnOnce(io::Error) + Clone + Send + 'static,
     {
-        let shutdown_signal = shutdown_signal.fuse();
-        let rt = tokio::runtime::Builder::new_multi_thread()
-            .enable_io()
-            .build()
-            .unwrap();
-
-        rt.block_on(async {
-            tokio::select! {
-                res = self.serve(service) => if let Err(e) = res { log::error!("Error: {}", e) },
-                _ = shutdown_signal => log::trace!("Shutdown signal received")
+        let abort_signal = abort_signal.fuse();
+        tokio::select! {
+            res = self.serve(on_connected, on_process_error) => {
+                res.map(|()| Terminated::Finished)
+            },
+            () = abort_signal => {
+                Ok(Terminated::Aborted)
             }
-        })
-    }
-
-    pub fn serve_forever<S, Req, Res>(self, service: S)
-    where
-        S: NewService<Request = Req, Response = Res> + Send + Sync + 'static,
-        Req: From<tcp::RequestAdu> + Send,
-        Res: Into<OptionalResponsePdu> + Send,
-        S::Instance: Send + Sync + 'static,
-        S::Error: Into<Error>,
-    {
-        self.serve_until(service, futures::future::pending())
+        }
     }
 }
 
 /// The request-response loop spawned by serve_until for each client
 async fn process<S, Req, Res>(
-    framed: Framed<TcpStream, codec::tcp::ServerCodec>,
+    mut framed: Framed<TcpStream, ServerCodec>,
     service: S,
 ) -> io::Result<()>
 where
     S: Service<Request = Req, Response = Res> + Send + Sync + 'static,
-    S::Request: From<tcp::RequestAdu> + Send,
+    S::Request: From<RequestAdu> + Send,
     S::Response: Into<OptionalResponsePdu> + Send,
-    S::Error: Into<Error>,
+    S::Error: Into<io::Error>,
 {
-    let mut framed = framed;
-
     loop {
-        let request = framed.next().await;
-
-        // tcp socket closed
-        if request.is_none() {
+        let Some(request) = framed.next().await.transpose()? else {
+            log::debug!("TCP socket has been closed");
             break;
-        }
+        };
 
-        let request = request.unwrap()?;
         let hdr = request.hdr;
-        let response: OptionalResponsePdu = service
+        let OptionalResponsePdu(Some(response_pdu)) = service
             .call(request.into())
             .await
             .map_err(Into::into)?
-            .into();
+            .into() else {
+                log::trace!("Sending no response for request {hdr:?}");
+                continue;
+        };
 
-        match response.0 {
-            Some(pdu) => {
-                framed.send(tcp::ResponseAdu { hdr, pdu }).await?;
-            }
-            None => {
-                log::debug!("No response for request {hdr:?}");
-            }
-        }
+        framed
+            .send(ResponseAdu {
+                hdr,
+                pdu: response_pdu,
+            })
+            .await?;
     }
+
     Ok(())
 }
 
@@ -167,7 +177,8 @@ fn configure_tcp(_workers: usize, _tcp: &Socket) -> io::Result<()> {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::server::Service;
+
+    use crate::{prelude::*, server::Service};
 
     use futures::future;
 
@@ -181,7 +192,7 @@ mod tests {
         impl Service for DummyService {
             type Request = Request;
             type Response = Response;
-            type Error = Error;
+            type Error = io::Error;
             type Future = future::Ready<Result<Self::Response, Self::Error>>;
 
             fn call(&self, _: Self::Request) -> Self::Future {

--- a/src/slave.rs
+++ b/src/slave.rs
@@ -41,9 +41,9 @@ impl Slave {
     /// connected Modbus TCP device, i.e. if not forwarded through
     /// a TCP/RTU gateway according to the unit identifier.
     ///
-    /// [MODBUS Messaging on TCP/IP Implementation Guide](http://www.modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf), page 23
-    /// "On TCP/IP, the MODBUS server is addressed using its IP address; therefore,
-    /// the MODBUS Unit Identifier is useless. The value 0xFF has to be used."
+    /// [Modbus Messaging on TCP/IP Implementation Guide](http://www.modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf), page 23
+    /// "On TCP/IP, the Modbus server is addressed using its IP address; therefore,
+    /// the Modbus Unit Identifier is useless. The value 0xFF has to be used."
     #[must_use]
     pub const fn tcp_device() -> Self {
         Slave(255)

